pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -737,15 +737,15 @@ public void authChallengeSuccessCallback(AuthData authChallenge,
                 // 2. an authentication refresh, in which case we need to refresh authenticationData
                 AuthenticationState authState = useOriginalAuthState ? originalAuthState : this.authState;
                 String newAuthRole = authState.getAuthRole();
+                AuthenticationDataSource newAuthDataSource = authState.getAuthDataSource();
 
-                // Refresh the auth data.
-                this.authenticationData = authState.getAuthDataSource();
-                if (log.isDebugEnabled()) {
-                    log.debug("[{}] Auth data refreshed for role={}", remoteAddress, this.authRole);
-                }
-
+                // Refresh the auth data and role.
                 if (!useOriginalAuthState) {
                     this.authRole = newAuthRole;
+                    this.authenticationData = newAuthDataSource;
+                } else {
+                    this.originalAuthData = newAuthDataSource;
+                    this.originalPrincipal = newAuthRole;
```

Comment:
My primary concern about this approach is that it doesn't address the fact that the `authenticationData` will never be updated after the initial handshake when a proxy is involved.

I documented some of the challenges here https://github.com/apache/pulsar/issues/19332.

I have been thinking it might make sense to get rid of the `originalAuthData` field entirely because when it is set, we won't ever get the proxy's auth data. The drawback of my approach is that it also means we would skip authorization of the proxy's role. I personally think this is reasonable because we can verify that the proxy is in the `proxyRoles` initially and then not worry about the actual permission later.

However, it could make sense to merge this PR and then implement my proposed change later. I will spend some time thinking about this and provide more feedback tomorrow.

Reply:
> My primary concern about this approach is that it doesn't address the fact that the `authenticationData` will never be updated after the initial handshake when a proxy is involved.

Sure, but I'm doing this to fix the store of the authentication data correctly, and make sense.

> I have been thinking it might make sense to get rid of the `originalAuthData` field entirely because when it is set, we won't ever get the proxy's auth data. The drawback of my approach is that it also means we would skip authorization of the proxy's role. I personally think this is reasonable because we can verify that the proxy is in the `proxyRoles` initially and then not worry about the actual permission later.

It makes sense to remove 'originalAuthData' to make the authentication process simpler.

Thank you for your review.

Reply:
> Sure, but I'm doing this to fix the store of the authentication data correctly, and make sense.

I think I was too focused on the end goal. Your solution is a good first step because it corrects how we store the authentication data. I think we should merge this before worrying about changing the whole paradigm.

Reply:
Simplifying the authentication process is an important thing, I also think is a good goal, but I've thought about it and haven't found a way to do it yet, so made this PR.

This PR does not affect the built-in plugins. 

After `AuthChallenge`, once the AuthorizationProvider checks the role and authentication data, which will find a mismatch, this happened in our project.

Reply:
> After `AuthChallenge`, once the AuthorizationProvider checks the role and authentication data, which will find a mismatch, this happened in our project.

Yes, that is exactly the issue I noticed while writing https://github.com/apache/pulsar/issues/19332. This is an important fix.